throttle/contract_tx_limiter_middleware.go

```diff
@@ -0,0 +1,178 @@
+}
+
+type contractTxLimiter struct {
+	// contract_address to limiting parametres structure
+	contractToTierMap       map[string]udw.TierID
+	contractDataLastUpdated int64
+	// track of no. of txns in previous blocks per contract
+	contractToBlockTrx  map[string]*blockTxn
+	tierMap             map[udw.TierID]udw.Tier
+	tierDataLastUpdated int64
+}
+
+type blockTxn struct {
+	txn         int64
+	blockHeight int64
+}
+
+var TxLimiter *contractTxLimiter
+
+func (txl *contractTxLimiter) isAccountLimitReached(contractAddr loom.Address, curBlockHeight int64) bool {
+	blockTx, ok := txl.contractToBlockTrx[contractAddr.String()]
+	if !ok {
+		return false
+	}
+	// if execution reaches here => tierID and tier are valid
+	tierID := txl.contractToTierMap[contractAddr.String()]
+	tier := txl.tierMap[tierID]
+	if blockTx.blockHeight <= curBlockHeight-int64(tier.BlockRange) || int64(tier.MaxTx) > blockTx.txn {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (txl *contractTxLimiter) updateState(contractAddr loom.Address, curBlockHeight int64) {
+	blockTx, ok := txl.contractToBlockTrx[contractAddr.String()]
+	tierID := txl.contractToTierMap[contractAddr.String()]
+	tier := txl.tierMap[tierID]
+	if !ok || blockTx.blockHeight <= curBlockHeight-int64(tier.BlockRange) {
+		// resetting the blockHeight to lower bound of range instead of curblockheight
+		rangeStart := (((curBlockHeight - 1) / int64(tier.BlockRange)) * int64(tier.BlockRange)) + 1
+		txl.contractToBlockTrx[contractAddr.String()] = &blockTxn{1, rangeStart}
+		return
+	}
+	blockTx.txn++
+}
+
+// NewContractTxLimiterMiddleware add another tx limiter that limits how many CallTx(s) can be sent to an EVM contract within a pre-configured block range
+func NewContractTxLimiterMiddleware(cfg *ContractTxLimiterConfig,
+	createUserDeployerWhitelistCtx func(state loomchain.State) (contractpb.Context, error),
+) loomchain.TxMiddlewareFunc {
+	TxLimiter = &contractTxLimiter{
+		contractToBlockTrx: make(map[string]*blockTxn, 0),
+	}
+	return loomchain.TxMiddlewareFunc(func(
+		state loomchain.State,
+		txBytes []byte,
+		next loomchain.TxHandlerFunc,
+		isCheckTx bool,
+	) (res loomchain.TxHandlerResult, err error) {
+		if !isCheckTx {
+			return next(state, txBytes, isCheckTx)
+		}
+		var nonceTx auth.NonceTx
+		if err := proto.Unmarshal(txBytes, &nonceTx); err != nil {
+			return res, errors.Wrap(err, "throttle: unwrap nonce Tx")
+		}
+		var tx loomchain.Transaction
+		if err := proto.Unmarshal(nonceTx.Inner, &tx); err != nil {
+			return res, errors.New("throttle: unmarshal tx")
+		}
+		if tx.Id != callId {
+			return next(state, txBytes, isCheckTx)
+		}
+		var msg vm.MessageTx
+		if err := proto.Unmarshal(tx.Data, &msg); err != nil {
+			return res, errors.Wrapf(err, "unmarshal message tx %v", tx.Data)
+		}
+		var msgTx vm.CallTx
+		if err := proto.Unmarshal(msg.Data, &msgTx); err != nil {
+			return res, errors.Wrapf(err, "unmarshal call tx %v", msg.Data)
+		}
+		if msgTx.VmType == vm.VMType_EVM {
```

Comment:
Get rid of one level of nesting by doing this:
```golang
if msgTx.VmType != vm.VMType_EVM {
  return next(state, txBytes, isCheckTx)
}
```

Reply:
done